R/PseudoBulk.R

```diff
@@ -352,3 +356,140 @@ CreateStudyWideBarPlot <- function(pairwise_de_results, pvalue_threshold = 0.05,
   
   return(DEG_bargraph)
 }
+
+#' @title FilterPseudobulkContrasts
+#'
+#' @description This is designed to accept a study design defined by a series of logical gates applied to metadata fields, test them for equivalence to some criterion, and then filter pairwise contrasts accordingly.
+#' @param logical_dataframe The dataframe that defines the study design. Please see examples for the setup and use. This dataframe should have three columns: field_to_check, field_logic, and criteria. field_to_check defines the metadata field (e.g. Tissue or Timepoint) to which the logic gate will be applied. field_logic defines the logic gate that will be used for that field (one of: all (AND), any (OR), xor, nand, nor, or xnor. criteria defines the specific value of the metadata field that will be tested for equivalence against the positive and negative contrasts within the gate.
+#' @param design a design/model matrix returned by DesignModelMatrix().
+#' @param contrast_columns The vector of contrast names supplied to DesignModelMatrix() upstream of this function. Please ensure this vector is identical to the contrast_columns vector supplied to DesignModelMatrix().
+#' @param use_require_identical_logic Whether or not to apply require_identical logic using require_identical_fields. It is possible to use logic gates to keep metadata fields constant, but using this saves time & effort/iteration.
+#' @param require_identical_fields The metadata columns of the SeuratObj that you wish to keep constant. For example: defining "Tissue" in this vector would filter all contrasts comparing Liver to PBMC, as their tissue is not identical.
+#' @param filtered_contrasts_output_file The file to write the list of filtered contrasts to (one pair of samples per row, separated into their metadata fields).
+#' @return A dataframe of pairwise contrasts, with column names indicating the directionality (positive vs negative) and associated metadata column (inferred by contrast_columns).
+#' @examples
+#' \dontrun{
+#' #set up design matrix
+#' contrast_columns = c("Timepoint", "Vaccine", "Challenge", "Tissue", "cell_type", "SampleType")
+#' design <- DesignModelMatrix(seuratObj, contrast_columns = c("Timepoint", "Vaccine", "Challenge", "Tissue", "cell_type","SampleType"), sampleIdCol = 'cDNA_ID')
+#' 
+#' 
+#' #form study design using logic gates. Each gate consists of a metadata field (field_to_check) to apply the logic gate to 
+#' gate_one <- data.frame(field_to_check = "Challenge", field_logic = "xor", criteria = "Mock.challenged")
+#' gate_two <- data.frame(field_to_check = "SampleType", field_logic = "all", criteria = "Necropsy")
+#' logical_dataframe <- rbind(gate_one, gate_two)
```

Comment:
Also good idea on reformatting the construction of the logical_dataframe. Now, the example constructs gates one at a time and rbinds them. I think this is close to what you suggested in Teams without needing to parse/coerce the list into a data frame.